packages/world/src/ResourceId.sol

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0;
+import { ROOT_NAMESPACE, ROOT_NAME } from "./constants.sol";
+import { Bytes } from "@latticexyz/store/src/Bytes.sol";
+import { ResourceType } from "@latticexyz/store/src/ResourceType.sol";
+import { RESOURCE_NAMESPACE, MASK_RESOURCE_NAMESPACE } from "./worldResourceTypes.sol";
+
+bytes16 constant ROOT_NAMESPACE_STRING = bytes16("ROOT_NAMESPACE");
+bytes16 constant ROOT_NAME_STRING = bytes16("ROOT_NAME");
+
+bytes32 constant NAMESPACE_MASK = bytes32(~bytes14(""));
+
+// TODO: change all namespace functions to namespace id functions (don't use bytes14 but bytes32 for ownership and access control)
+// TODO: should resource id be a user type to avoid type errors of eg passing a namespace (bytes14) instead of a namespace id (bytes32)
+// (I ran into a bunch of runtime errors while refactoring this where bytes14 was automatically casted to bytes32)
+
+library ResourceId {
+  /**
+   * Create a 32-byte resource ID from a namespace, name and type.
+   *
+   * A resource ID is a 32-byte value that uniquely identifies a resource.
+   * The first 14 bytes represent the namespace,
+   * the next 16 bytes represent the name,
+   * the last 2 bytes represent the type.
```

Comment:
ooc why last two bytes? in terms of information hierarchy, type feels like it should be first

Reply:
we check the namespace more often and in hotter paths (in access control) and it's cheaper to check the most significant bits because we don't need to shift

Reply:
does that mean type is least often used compared to name?

Reply:
i'd say `type` and `name` are used similarly often (in access control checks, and in the future in "type" checks if logic depends on it). I did some quick benchmarks and it seems like it's not significant how much you shift, just whether or not you shift, so it doesn't really matter if `name` or `type` would come first